### PR TITLE
Breton/apply effective kotlin

### PR DESCRIPTION
**Pull request by noterB**

### CODE REVIEW

File: tests/snapshots/FeatureAPI.kt [hunk 1]
```diff
@@ -69,40 +69,193 @@ class FeatureAPI : GBIORxWebServices() {
...
+        override fun toString() = "FeaturesFeatures(n=$n, r=$r, v=$v, p=$p, l=$l)"
+        
+        override fun hashCode(): Int {
+            var result:Int = n.hashCode()
+            result = 31 * result + r.hashCode()       
+            result = 31 * result + v.hashCode()       
+            result = 31 * result + p.hashCode()       
+            result = 31 * result + l.hashCode()       
+            return result
+        }
```
Comment (prashermanas) [inline tests/snapshots/FeatureAPI.kt]: Can we not have a simpler version of hashcode Java gives us the `Objects.hash` method which is easier to read too

`Objects.hash(n.hashCode(), r.hashCode() ,v.hashCode(),p.hashCode(),l.hashCode())`

Comment (fergusonm) [reply tests/snapshots/FeatureAPI.kt]: This looks like a variant of what the IDE would generate for hashcode and equals.  Looks ok to me.

Comment (noterB) [reply tests/snapshots/FeatureAPI.kt]: I suppose so, I just cribed the IDE generated one. Its all generated so i'm not really concerned about readability.

File: tests/snapshots/FeatureAPI.kt [hunk 2]
```diff
@@ -69,40 +69,193 @@ class FeatureAPI : GBIORxWebServices() {
...
+                override fun hashCode(): Int {
+                    var result:Int = id.hashCode()
+                    result = 31 * result + os.hashCode()       
+                    result = 31 * result + ver.hashCode()       
+                    return result
+                }
+            
+                override fun equals(other: Any?): Boolean {
+                    if (this === other) return true
+                    if (javaClass != other?.javaClass) return false
```
Comment (fergusonm) [inline tests/snapshots/FeatureAPI.kt]: So this breaks content equality for subclasses that have the same content as a super class.  I'm fine with that, just double checking intentions.

Comment (fergusonm) [reply tests/snapshots/FeatureAPI.kt]: Yeah I'm ok with that.

**Pull request merged**